### PR TITLE
28 is leap year

### PR DESCRIPTION
**Pull request by Khusro-S**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/is_leap_year.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Module contents:
+    - is_leap: determines whether the given year is leap year or not.
+
+Created on XX XX XX
+@author: Khusro Sakhi
+"""
+
+
+def is_leap_year(year) -> bool:
```
Comment (Mohamed-Elnageeb) [inline solutions/is_leap_year.py:14]: you forgot inputs type annotations

File: solutions/tests/test_is_leap_year.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+"""
+Test module for is_leap function.
+Contains intentionally buggy tests for debugging practice.
+
+Test categories:
+    - Standard cases: Common leap and non-leap years
+    - Edge cases: Smallest leap year, non-leap years close to leap years
+    - Defensive tests: Handling of invalid inputs (e.g., non-integer, negative values)
+
+Created on XX XX XX
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_is_leap_year.py:12]: add date

File: solutions/is_leap_year.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    False
+
+    >>> is_leap(2100)
+    False
+
+    >>> is_leap(2004)
+    True
+    """
+    assert isinstance(year, int), "entered year is not an integer"
+    assert year >= 0, "year is less than 0"
```
Comment (Mohamed-Elnageeb) [inline solutions/is_leap_year.py:42]: i dont know but is there a year zero? 
if not then this should be year > 0.

**Review by Mohamed-Elnageeb — COMMENTED** [review]
Nice solution! I appreciate how compact it is, although some code norms might consider it less readable. However, I think it works well. The tests pass all the checklist requirements and are thorough and well-rounded. Great job! I have only highlighted a few small changes.

**Review by YonatanBest — DISMISSED** [review]
Great job! They passed all the ci checks, and everything seems correct for me. But could you change Created on XX XX XX to specific date?

File: solutions/is_leap_year.py (hunk 1 above)
Comment (Khusro-S) [reply solutions/is_leap_year.py:14]: You're right I happen to have missed it 😅.

File: solutions/tests/test_is_leap_year.py (hunk 2 above)
Comment (Khusro-S) [reply solutions/tests/test_is_leap_year.py:12]: @YonatanBest also mentioned this, I'll update it on my other challenge as well 👍

File: solutions/is_leap_year.py (hunk 3 above)
Comment (Khusro-S) [reply solutions/is_leap_year.py:42]: Good catch, I searched about it and turns out year 0 is widely not accepted as a valid year 👍

**Comment by Khusro-S** [comment]
I'll make these corrections soon, thanks for the feedback 🙌

**Comment by Khusro-S** [comment]
Thank you both @YonatanBest and @Mohamed-Elnageeb. I made all the suggested changes and corrections 🙌.
Kindly review the changes one last time to make sure I didnt miss anything this time 😅 .

**Review by YonatanBest — APPROVED** [review]
They passed all the ci checks. And They are well written. Great Job!

**Comment by Khusro-S** [comment]
> They passed all the ci checks. And They are well written. Great Job!

Thank you 🙌

**Review by Mohamed-Elnageeb — APPROVED** [review]
I see no changes needed. Good job!

**Review by Mohamed-Elnageeb — APPROVED** [review]
I see no changes needed. Good job!

**Review by Zeinab15 — COMMENTED** [review]
Well done @Khusro-S , just a few suggestions

File: solutions/is_leap_year.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    Returns:
+        bool: True if the year is a leap year, False otherwise.
+
+     Raises:
+        AssertionError: If the input is a string.
+        AssertionError: If the input is a negative year.
+        AssertionError: If the input is None.
+        AssertionError: If the input is a float.
+
```
Comment (Zeinab15) [inline solutions/is_leap_year.py:28]: Replace the generic AssertionError with specific error types for better clarity as:
TypeError: Raised if the input is not an integer (e.g., a string or float).
ValueError: Raised if the year is less than or equal to 0.

and edit your code to:
if not isinstance(year, int):
    raise TypeError("TypeError: The input must be an integer. You provided a non-integer value (e.g., string, float).")

if year <= 0:
    raise ValueError("ValueError: The year must be a positive integer. Provided year is less than or equal to 0.")

Comment (Zeinab15) [reply solutions/is_leap_year.py:28]: This update will improve the code's readability, provide clearer feedback during execution, and help users or developers quickly identify and fix errors.

File: solutions/tests/test_is_leap_year.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    def test_none_input(self):
+        """It should raise AssertionError for None input"""
+        with self.assertRaises(AssertionError):
+            is_leap_year(None)
+
+    def test_float_input(self):
+        """It should raise AssertionError for a float input"""
+        with self.assertRaises(AssertionError):
+            is_leap_year(4.5)
```
Comment (Zeinab15) [inline solutions/tests/test_is_leap_year.py:73]: Could you add another edge case test that raises a ValueError, such as:
def test_zero_year(self):
    """It should raise ValueError for year 0, as year 0 is not valid in the Gregorian calendar."""
    with self.assertRaises(ValueError):
        is_leap_year(0)

**Review by Zeinab15 — COMMENTED** [review]
The code is well-structured with good test coverage, but there are a few areas for improvement:

Error Handling: Replace AssertionError with more specific exceptions like TypeError for invalid types and ValueError for non-positive years.
Test Coverage: Consider adding a test for the year 0 to raise a ValueError, as it's an invalid input.

**Comment by Khusro-S** [comment]
> The code is well-structured with good test coverage, but there are a few areas for improvement:
> 
> Error Handling: Replace AssertionError with more specific exceptions like TypeError for invalid types and ValueError for non-positive years. Test Coverage: Consider adding a test for the year 0 to raise a ValueError, as it's an invalid input.

Yeah, you’re right, that’s a good point and definitely best practice. I’ll make those changes in a bit. Thank you 🎉

**Review by Zeinab15 — APPROVED** [review]
Good job @Khusro-S

**Review by Mohamed-Elnageeb — APPROVED** [review]

File: solutions/is_leap_year.py (hunk 4 above)
Comment (YonatanBest) [reply solutions/is_leap_year.py:26]: Great!

**Review by YonatanBest — APPROVED** [review]
Great job!
